### PR TITLE
feature/implement_uidefinition_token_workflow

### PR DESCRIPTION
**Pull request by Blokh**
feat: integrated token plugin, ui definition, workflow runtime, endpoints

## Summary by CodeRabbit

- **New Features**
	- Introduced new types and interfaces for enhanced plugin functionality, including `TWorkflowTokenPluginCallback` and `ISerializableWorkflowTokenPlugin`.
	- Added `WorkflowTokenPlugin` class for managing workflow token actions.
	- Enhanced `WorkflowRunner` and `WorkflowService` for processing workflow token actions, including error handling and data retrieval.
	- Extended database schema to support UI definitions linked to workflow runtime data.
	- Added flexible querying capabilities to UI definition methods by including optional parameters.

- **Bug Fixes**
	- Improved error handling in workflow token action management.

### CODE REVIEW

File: services/workflows-service/src/ui-definition/ui-definition.repository.ts [hunk 1]
```diff
@@ -40,11 +40,16 @@ export class UiDefinitionRepository {
     uiContext: keyof typeof UiDefinitionContext,
     args: Prisma.SelectSubset<T, Omit<Prisma.UiDefinitionFindFirstOrThrowArgs, 'where'>>,
```
Comment (liorzam) [inline services/workflows-service/src/ui-definition/ui-definition.repository.ts]: you can use args for fetch by id, just edit to have to pick the id property from the where prima arg object

Comment (Blokh) [reply services/workflows-service/src/ui-definition/ui-definition.repository.ts]: mit<Prisma.UiDefinitionFindFirstOrThrowArgs, 'where'> means, without where...
I can change it to include where but then I'll have to add where everythwere in the code

File: services/workflows-service/prisma/schema.prisma [hunk 2]
```diff
@@ -429,9 +435,12 @@ model UiDefinition {
   createdAt     DateTime  @default(now())
   updatedAt     DateTime? @updatedAt
 
-  projectId          String
-  project            Project            @relation(fields: [projectId], references: [id])
-  workflowDefinition WorkflowDefinition @relation(fields: [workflowDefinitionId], references: [id])
+  projectId             String
+  project               Project               @relation(fields: [projectId], references: [id])
+  workflowDefinition    WorkflowDefinition    @relation(fields: [workflowDefinitionId], references: [id])
+  workflowRuntimeDatass WorkflowRuntimeData[]
```
Comment (alonp99) [inline services/workflows-service/prisma/schema.prisma]: data ass?

File: services/workflows-service/src/ui-definition/ui-definition.repository.ts [hunk 3]
```diff
@@ -33,18 +33,26 @@ export class UiDefinitionRepository {
...
-  async findByWorkflowDefinitionId<
-    T extends Omit<Prisma.UiDefinitionFindFirstOrThrowArgs, 'where'>,
-  >(
+  async findByArgs(args: Prisma.UiDefinitionFindFirstOrThrowArgs, projectIds: TProjectIds) {
+    return await this.prisma.uiDefinition.findFirstOrThrow(
+      this.scopeService.scopeFindFirst(args, projectIds),
+    );
+  }
+
+  async findByWorkflowDefinitionId(
```
Comment (alonp99) [inline services/workflows-service/src/ui-definition/ui-definition.repository.ts:42]: findByWorkflowDefinitionId?

File: services/workflows-service/src/workflow/workflow.service.ts [hunk 4]
```diff
@@ -2000,6 +2001,81 @@ export class WorkflowService {
             transaction,
           );
         },
+        invokeWorkflowTokenAction: async (workflowTokenAction: TWorkflowTokenPluginCallback) => {
+          const workflowRuntimeId = workflowTokenAction.workflowRuntimeId;
+          const defaultDaysExpiry = 30;
+
+          const expiresAt = workflowTokenAction.expiresInMinutes
+            ? new Date(Date.now() + workflowTokenAction.expiresInMinutes * 60 * 1000)
+            : new Date(Date.now() + defaultDaysExpiry * 24 * 60 * 60 * 1000);
```
Comment (alonp99) [inline services/workflows-service/src/workflow/workflow.service.ts:2010]: I think you confused between days and minutes

Comment (Blokh) [reply services/workflows-service/src/workflow/workflow.service.ts:2010]: the default is days 30 days ( 24 * 60 )  - in days
and 60 (seconds) * 1000 ( milliseconds) should be well configured

File: packages/workflow-core/src/lib/workflow-runner.ts [hunk 5]
```diff
@@ -741,10 +768,20 @@ export class WorkflowRunner {
     }
 
     if (!!response) {
-      this.context.pluginsOutput = {
-        ...(this.context.pluginsOutput || {}),
-        ...{ [commonPlugin.name]: response },
-      };
+      if (hasPersistResponseDestination(commonPlugin)) {
+        if (response) {
```
Comment (alonp99) [inline packages/workflow-core/src/lib/workflow-runner.ts:773]: Seems like its redundant

File: services/workflows-service/prisma/migrations/20240731131954_add_ui_definition_naming/migration.sql [hunk 6]
```diff
@@ -0,0 +1,6 @@
+-- AlterTable
+ALTER TABLE "UiDefinition" ADD COLUMN     "displayName" TEXT,
+ADD COLUMN     "name" TEXT;
+
+-- CreateIndex
+CREATE INDEX "UiDefinition_name_projectId_idx" ON "UiDefinition"("name", "projectId");
```
Comment (alonp99) [inline services/workflows-service/prisma/migrations/20240731131954_add_ui_definition_naming/migration.sql:6]: didnt you say you wanted to add it afterwards? what would happen with empty names in current envs?

Comment (Blokh) [reply services/workflows-service/prisma/migrations/20240731131954_add_ui_definition_naming/migration.sql:6]: it's an index not unique index, it's for searching, etc. later on we'll alter and make it unique index

File: packages/workflow-core/src/lib/workflow-runner.ts (hunk 5 above)
Comment (Blokh) [reply packages/workflow-core/src/lib/workflow-runner.ts:773]: you receive either response or error.
we need to check if the response is appearing and not the error

**Pull request merged**